### ISSUE TITLE
Add SIF forward operator for CLM and land modeling

### ISSUE THREAD

**Issue by braczka** [NCAR/DART#338]
### Use case  
My goal is to add both SIF observation waveband conversion and provide multiple approaches to calculate the expected SIF
within the DART forward operator.  The DART code has already laid some groundwork for a SIF forward operator.  For example
 an observation converter was added in spring 2021 for ``HARMONIZED_SIF`` within ``~/observations/obs_converters/SIF``. Furthermore, ``read``, ``write``, and ``interactive_SIF_wavelength`` scripts were added to ``obs_def_land_mod.f90`` in anticipation that a waveband converter would be added at some point.   

### Is your feature request related to a problem?  
Currently, no wavelength conversion exists for the SIF forward operator -- the ``SIF_wavelength`` scripts within ``obs_def_land_mod.f90`` are simply pass-through, thus the SIF forward operator acts as a common code spatial interpolation of the SIF quantity state variable.   

A separate challenge is that representing SIF is still a work in progress for the earth system modeling community, thus model simulated SIF is subject to significant systemic errors.  For example CLM model simulations of SIF show relatively high skill for deciduous and grass biomes, but poor skill for evergreen needleleaf biomes as described [here]( https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2021MS002747).   These large errors increase the likelihood that valid observations of SIF will be rejected by the outlier threshold within the DART assimilation framework.   This loss of information could be regained if we employ an alternative, empirical approach to construct the SIF forward operator.  For example SIF is known to have a strong biome specific linear relationship  with GPP (MacBean et al., 2018;  Bacour, Sun et al., 2019).  Thus by constructing a linear forward operator based on GPP, it could still be possible to leverage SIF information during an assimilation.

### Describe your preferred solution  
First, I intend to add waveband conversion code within  the``obs_def_land_mod.f90`` ``SIF_wavelength`` scripts to properly convert between SIF wavebands (e.g. SIF can be reported at 740 or 757 nm) such that there is no mismatch in wavelength between the expected and observed SIF.

Second, I intend to provide the DART user a choice between a mechanistic or empirical SIF forward operator based upon the availability of SIF and GPP quantities within the ``&model_nml``.   The mechanistic forward operator will rely on an external model (like CLM) to provide the SIF state through the quantity ``QTY_SOLAR_INDUCED_FLUORESCENCE``.   Thus if a SIF quantity is provided for ``model_mod.f90`` through the  ``&model_nml`` that quantity will be spatially interpolated  and then a wavelength conversion will be applied to create the expected SIF.   

Alternatively, the empirical forward operator will be applied when the SIF quantity is intentionally withheld and instead the GPP quantity (``QTY_GROSS_PRIMARY_PROD_FLUX``) is provided  through the ``&model_nml``.     Currently, the interpolate script within ``model_mod.f90`` for CLM spatially weights the biome specific GPP values to calculated the grid cell average GPP (expected GPP).    To calculate the grid cell average SIF, I intend to add a biome-specific GPP-SIF linear operator during this weighting process.   This operation requires biome specific metadata, thus for CLM, this requires the GPP state be provided as a ``vector_history`` file.   As before, the waveband conversion will then be applied in order to match the SIF wavelength of the observation.

### Describe any alternatives you have considered  
My approach for providing a mechanistic/empirical SIF forward operator based upon the availability of specific quantities within ``&model_nml`` loosely follows a previous approach for assimilating soil moisture observations.   For soil moisture DA the CLM model_mod will first look for a soil moisture quantity within the model state to spatially interpolate.  If that quantity does not exist it will default to constructing the forward operator through several alternative quantities.   Thus my approach, to some extent, follows precedent.

My proposal for constructing the empirical SIF forward operator is necessarily applied within CLM's ``model_mod.f90`` because the biome specific SIF-GPP operator requires model specific biome metadata.  Similar ``model_mod`` code would have to be added to perform SIF assimilation with a land model other than CLM.   This seems redundant, but I am not sure there is a workaround.   The waveband conversion code, on the other hand, I intend to add to ``obs_def_land_mod.f90`` which is visible to all models (not just CLM).

**Comment by hkershaw-brown** [comment]
Hi Brett, what do you think about splitting obs_def_land_mod.90 so the SIFs fwd operator has its own file?

obs_def_SIF_mod.f90 
- gives you a separate namelist (maybe the wavelength options, maybe force an empircal calcluation)

I'm looking at the other fwd operators. Sometimes grouped by land, cice, ocean, etc. sometimes they are grouped by observation (e.g. GRACE).  If we break out SIF into its own obs_def that can make it easier for people to switch out their own SIF fwd operator (I'm assuming there will be a lot of active development on the best way to calcualte SIF).

**Comment by braczka** [comment]
Hi Helen --  I am OK with splitting SIFs forward operator out from obs_def_land_mod.f90 into an obs_def_SIF_mod.f90.   However, I am not sure how much future development will go into the obs_def_SIF portion of the code.  I plan to do the wavelength conversion there (typically it will shift between the SIF peak emission bands, between 740 nm and 757 nm or vice versa).  Most of the heavy lifting for the SIF forward operator I intend to insert within the CLM model_mod code.   This is where I plan to insert the biome specific linear operator to convert from GPP to SIF.   This is where user's might want to insert a different operator of their own choosing.  I haven't decided yet what manuscript I will pull from to supply the coefficients for this conversion.       So - yes, I am OK with splitting up the obs_def_land and obs_def_SIF to keep the obs_def_land uncluttered,  but the only development in the obs_def_SIF might be the wavelength converter part of the forward operator.   Let me know what you think.

**Comment by hkershaw-brown** [comment]
Keeping in obs_def_land_mod.f90 seems like the plan then.  I figured I'd just throw splitting obs_def out there before you got too far along. Looking forward to this biome stuff.

**Comment by braczka** [comment]
Cool -- sounds like we are on the same page.  The biome specific nature of the forward operator here will be specific to the SIF forward operator.  However, this reminds me that something similar could be done to apply biome metadata to land surface observations  in general to create a universal application of biome metadata to some land-based forward operators.   For example, for biomass or leaf area observations we could bring in biome metadata with these observations, and purposely filter out all non-similar biomes when constructing the expected biomass or leaf area from the model state.  If the biome metadata is not available for certain land-based observations, we could infer it using the observation location and high resolution biome maps -- probably the same one that the CLM group uses to create land surface maps.   I think this SIF issue will help me get more comfortable with the forward operator code, before taking that on that larger goal.

**Comment by nancycollins** [comment]
hi brett - here's some philosophical background on forward operator design.  if you need to do it differently at least you know what the intent was.

we prefer the model_mod only supports returning interpolated values for fields in the state.  any forward operator that needs multiple values and computes on them should be in an obs_def module.  that way it can be used by any model of the same type (land, atm, etc).  the obs_def code can call interpolate multiple times for the fundamental values it needs and then compute the expected value for each ensemble member.

there are already exceptions - usually justified by some kind of efficiency argument.  for example, there is a lot of code in the WRF model_mod to handle vortex centers, etc.  that code really should be in an obs_def.  it could call interpolate for a grid of model points around the vortex center and then compute values.  that way it would work for mpas and cam and any other atmospheric model.  if you look at the overall run time, the forward operator code time is negligible so the efficiency argument is often specious.

the other thing people get stumped by is asking for the grid size.  often it's the number of vertical layers in the model.  several forward operators first ask for something in the state, like pressure or moisture, for artificial locations using VERTISLEVEL for 1,2,3 ... and loop until the interpolate call fails.  that's an easy way to count the levels and then the rest of the obs_def code can use that count for other computations.

i don't know any of the details about the SIF obs, but if there's any way to put the calculations into a model-independent format and only ask the model for the needed input values to compute it, that would fit with the dart philosophy very well.

**Comment by nancycollins** [comment]
there aren't other land models to share this code, and also some kind of support would be needed to retrieve biome information in order to make a model-independent version.  so putting the code into the model_mod seems like the right call here.  (we discussed this in the standup - this note/comment is here to keep a record of why the decision was made.)

**Comment by braczka** [comment]
Here are some updates to progress on this issue -- some of this is documenting what has been said in stand-ups the past few weeks: 

 First, the mechanistic SIF forward operator approach requires the addition of SIF sourcemods within CLM such that it can produce a 'canopy SIF' variable for DART to calculate the expected SIF observation.   I can add the Sourcemods manually to the CLM-DART Sourcemods, *or* use the upcoming SIF branch that the CGD group is currently working on.   The issue for this work is here:  https://github.com/ESCOMP/CTSM/issues/1672.  CGD plans to convert this into a PR such that the SIF code can be accessed through a branch through CTSM, then eventually, they will create another PR to update the code to the very latest CTSM version.   To run SIF assimilation with DART it would be much more efficient to check out the SIF branch of CTSM instead of cramming all the SIF changes to the CLM-DART Sourcemods. I am tracking CGD's progress on this.


Second -- for the empirical SIF approach, I have completed a fairly lengthy literature review to better understand the current parameterization for the linear GPP-SIF relationship.  The literature has expanded to include GPP-SIF parameterizations for the following SIF observations data sets:  OCO-2 (Li et al., 2018),  TROPOMI (Turner et al.., 2021) and CSIF/GOSIF reanalysis SIF products (Gao et al., 2021).   As expected, the GPP-SIF relationship is biome specific -- which I have anticipated and laid out a plan of how to incorporate this.   The parameters are also satellite/retrieval technique dependent, which will mean that we will likely have to develop several SIF observation types to accommodate the differences in parameterizations.   Thus far, I have created just 1 SIF observation type 'SIF_harmony' based upon the GOME-2/SCIAMACHY satellite.   This approach of multiple SIF observation types will require a look-up table or dictionary approach.    
This is somewhat similar to the RTTOV FO approach, thus Helen has encouraged me to review the documentation/code for RTTOV such that I can migrate anything useful to the SIF FO. 

Finally, in order to prepare for the SIF wavelength conversions I have been analyzing SCOPE model output.  SCOPE is a canopy radiation model commonly used for SIF simulations. SCOPE  simulates the complete SIF spectrum (650-850 nm) thus can be used as a conversion tool between specific SIF far-red wavebands -- most commonly 740, 743-758,  757 and 771.

**Comment by braczka** [comment]
Update on mechanistic SIF forward operator.    I have adapted the cesm2.1 SourceMod changes developed by the University of Virginia  (UVA) group to the cesm2.2 version compatible with DART.   I identified and fixed a parameter bug within the original code.   I compiled and tested the SIF SourceMod changes within a global CLM free run and the year ran smoothly.    I am currently benchmarking the SIF behavior at a single site (Niwot Ridge) against the UVA group simulation.  Once that benchmarking is complete I will update the CLM-DART SourceMods to include these SIF changes.

The empirical SIF approach will go on the backburner for now.
